[popover] Implement popover focusing steps

nt1m opened the pull request:
#### 3558eb52e9aca998f3b1319d3f53dfe1ce40c931
<pre>
[popover] Implement popover focusing steps
<a href="https://bugs.webkit.org/show_bug.cgi?id=253231">https://bugs.webkit.org/show_bug.cgi?id=253231</a>
rdar://106132025

Reviewed by Ryosuke Niwa.

<a href="https://html.spec.whatwg.org/#popover-focusing-steps">https://html.spec.whatwg.org/#popover-focusing-steps</a>

PR open here to update the popover focusing steps: <a href="https://github.com/whatwg/html/pull/8998">https://github.com/whatwg/html/pull/8998</a>

The remaining failures are due to new dialog focusing steps being not implemented yet (TBD in <a href="https://bugs.webkit.org/show_bug.cgi?id=250795)">https://bugs.webkit.org/show_bug.cgi?id=250795)</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-child-dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::autoFocusDelegate):
(WebCore::Element::findFocusDelegateForTarget):
(WebCore::Element::findAutofocusDelegate):
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::runPopoverFocusingSteps):
(WebCore::HTMLElement::showPopover):

Canonical link: <a href="https://commits.webkit.org/261400@main">https://commits.webkit.org/261400@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f47dda03f4c1337a717247f4cb9accf4bc425357

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111616 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20750 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/222 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3370 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120362 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115676 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22106 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11830 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117381 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/99565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104476 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/143 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/45263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13234 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/140 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/87704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13726 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19170 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52132 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7938 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15712 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1284,6 +1284,25 @@ static HTMLElement* topmostPopoverAncestor(Element& newPopover)
+    if (!control)
+        return;
+
+    if (control->isFocusable())
```
I am not sure we should be checking for `isFocusable`.

Was reading through the focusing steps and step 1 says:

> If new focus target is not a [focusable area](https://html.spec.whatwg.org/#focusable-area), 
> then set new focus target to the result of [getting the focusable area](https://html.spec.whatwg.org/#get-the-focusable-area) 
> for new focus target, given focus trigger if it was passed.

nt1m replied:
@rniwa Do we implement "getting the focusable area" somewhere? I'm trying to implement https://html.spec.whatwg.org/#popover-focusing-steps

rniwa replied:
Not really. The closest we have is `Element::focusAppearanceUpdateTarget()`.

nt1m replied:
I removed the check for `isFocusable()`, we should fix `runFocusingStepsForAutofocus` separately to use `focusAppearanceUpdateTarget` if needed.

stwrt replied:
That sounds good to me.

stwrt commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1339,14 +1358,14 @@ ExceptionOr<void> HTMLElement::showPopover()
         { CSSSelector::PseudoClassOpen, true },
         { CSSSelector::PseudoClassClosed, false }
     });
     popoverData()->setVisibilityState(PopoverVisibilityState::Showing);
```
If the popover focusing steps fail should we be setting the visibility state still to showing?

nt1m replied:
The focusing steps should run after the setting the visibility state to showing (the FIXME comment was wrongly placed). See https://html.spec.whatwg.org/#show-popover

stwrt replied:
Ah, you are right; just did a read through of that piece of the spec this morning.

rniwa commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1284,6 +1284,27 @@ static HTMLElement* topmostPopoverAncestor(Element& newPopover)
+
+    control->runFocusingStepsForAutofocus();
+
+    if (!control->document().isFullyActive())
```
Where did this isFullyActive check come from? I don't see it in the spec.

nt1m replied:
removed

nt1m committed:
[popover] Implement popover focusing steps
https://bugs.webkit.org/show_bug.cgi?id=253231
rdar://106132025

Reviewed by Ryosuke Niwa.

https://html.spec.whatwg.org/#popover-focusing-steps

PR open here to update the popover focusing steps: https://github.com/whatwg/html/pull/8998

The remaining failures are due to new dialog focusing steps being not implemented yet (TBD in https://bugs.webkit.org/show_bug.cgi?id=250795)

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-2-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-child-dialog-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::autoFocusDelegate):
(WebCore::Element::findFocusDelegateForTarget):
(WebCore::Element::findAutofocusDelegate):
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::runPopoverFocusingSteps):
(WebCore::HTMLElement::showPopover):

Canonical link: https://commits.webkit.org/261400@main
